### PR TITLE
Support for vertex and hyperedge values

### PR DESCRIPTION
**Pull request by pszufe**
Support for vertex and hyperedge values following the discussion at issue #2 
This also includes several new tests and a bug correction for function `set_index!`

### CODE REVIEW

File: src/hypergraph.jl [hunk 1]
```diff
@@ -14,35 +14,61 @@ A hypergraph storing information about vertices and hyperedges.
...
 * `m` : a matrix representation rows are vertices and columns are hyperedges
 
 """
-struct Hypergraph{T} <: AbstractMatrix{Union{T, Nothing}}
+struct Hypergraph{T,V,E} <: AbstractMatrix{Union{T, Nothing}}
     v2he::Vector{Dict{Int,T}}
     he2v::Vector{Dict{Int,T}}
-    Hypergraph{T}(n, k) where {T<:Real} =
-        new([Dict{Int,T}() for i in 1:n], [Dict{Int,T}() for i in 1:k])
+    vs::Vector{V}
```
Comment (bkamins) [inline src/hypergraph.jl]: maybe beter `v_meta`  and `he_meta`? Not a crucial thing.

File: src/hypergraph.jl [hunk 2]
```diff
@@ -14,35 +14,61 @@ A hypergraph storing information about vertices and hyperedges.
...
+    hes::Vector{E}
+    Hypergraph{T,V,E}(n, k) where {T<:Real, V, E} =
+        new{T,V,E}([Dict{Int,T}() for i in 1:n], [Dict{Int,T}() for i in 1:k],Vector{V}(undef,n),Vector{E}(undef,k))
```
Comment (bkamins) [inline src/hypergraph.jl]:
```suggestion
        new{T,V,E}([Dict{Int,T}() for i in 1:n], [Dict{Int,T}() for i in 1:k],
                          Vector{V}(undef, n), Vector{E}(undef, k))
```

File: src/hypergraph.jl [hunk 3]
```diff
@@ -14,35 +14,61 @@ A hypergraph storing information about vertices and hyperedges.
...
 
 """
-struct Hypergraph{T} <: AbstractMatrix{Union{T, Nothing}}
+struct Hypergraph{T,V,E} <: AbstractMatrix{Union{T, Nothing}}
     v2he::Vector{Dict{Int,T}}
     he2v::Vector{Dict{Int,T}}
-    Hypergraph{T}(n, k) where {T<:Real} =
-        new([Dict{Int,T}() for i in 1:n], [Dict{Int,T}() for i in 1:k])
+    vs::Vector{V}
+    hes::Vector{E}
```
Comment (bkamins) [inline src/hypergraph.jl]: Looking at it maybe it would be better to make them `Vector{Union{Nothing, E}` and initialize to `nothing`. In this way we can know when metadata is set and when it is not set. Now you initialize it to garbage, but garbage is indistinguishable from a valid value for immutable types.

File: src/hypergraph.jl [hunk 4]
```diff
@@ -72,8 +98,8 @@ Removes a vertex from a given hyperedge for a hypergraph `h` and a given vertex-
 """
```
Comment (bkamins) [inline src/hypergraph.jl]: so maybe add a comment that trying to remove a vertex from a hyperedge when it is not present is not an error, earlier we wanted it to throw an error.

File: src/hypergraph.jl [hunk 5]
```diff
@@ -108,41 +134,100 @@ Returns hyperedges for a given vertex `v_id` in a hypergraph `h`.
...
 
 """
-    add_vertex!(h::Hypergraph{T};hyperedges::Dict{Int,T} = Dict{Int,T}()) where T <: Real
+    add_vertex!(h::Hypergraph{T, V, E}; hyperedges::Dict{Int,T} = Dict{Int,T}(), vertex_val::Union{UndefInitializer, V} = undef) where {T <: Real, V, E}
 
 Adds a vertex to a given hypergraph `h`. Optionally, the vertex can be added
 to existing hyperedges. The `hyperedges` parameter presents a dictionary
-of hyperedge identifiers and values stored at the hyperedges
+of hyperedge identifiers and values stored at the hyperedges.
+Additionally, a value can be stored with the vertex using the `vertex_val` parameter.
```
Comment (bkamins) [inline src/hypergraph.jl]: keyword parameter

File: src/hypergraph.jl (hunk 1 above)
Comment (pszufe) [reply src/hypergraph.jl]: One argument for `vs` is shorter code. I am not sure whether `_meta` is going to increase readibility. 
@spagnuolocarmine what do you think (since you proposed issue #2)

File: src/hypergraph.jl [hunk 6]
```diff
@@ -108,41 +134,100 @@ Returns hyperedges for a given vertex `v_id` in a hypergraph `h`.
...
 
 """
-function add_vertex!(h::Hypergraph{T};hyperedges::Dict{Int,T} = Dict{Int,T}()) where T <: Real
+function add_vertex!(h::Hypergraph{T, V, E}; hyperedges::Dict{Int,T} = Dict{Int,T}(), vertex_val::Union{UndefInitializer, V} = undef) where {T <: Real, V, E}
```
Comment (bkamins) [inline src/hypergraph.jl]: 1. make a line break
2. using `undef` here is a bad idea as its purpose in Julia is different. I would prefer `nothing`, and use the union with `Nothing` approach I have proposed above.

File: src/hypergraph.jl [hunk 7]
```diff
@@ -108,41 +134,100 @@ Returns hyperedges for a given vertex `v_id` in a hypergraph `h`.
...
     @boundscheck (checkbounds(h,1,k) for k in keys(hyperedges))
     push!(h.v2he,hyperedges)
     ix = length(h.v2he)
     for k in keys(hyperedges)
         h[ix,k]=hyperedges[k]
     end
+    if vertex_val == undef
+        resize!(h.vs, length(h.vs)+1)
+    else
+        push!(h.vs, vertex_val)
```
Comment (bkamins) [inline src/hypergraph.jl]: with `nothing` approach you can just push the `vertex_val` unconditionally.

File: src/hypergraph.jl [hunk 8]
```diff
@@ -108,41 +134,100 @@ Returns hyperedges for a given vertex `v_id` in a hypergraph `h`.
...
+    end
     ix
 end
 
 """
-    add_hyperedge!(h::Hypergraph{T};vertices::Dict{Int,T} = Dict{Int,T}()) where T <: Real
+    add_hyperedge!(h::Hypergraph{T, V, E}; vertices::Dict{Int,T} = Dict{Int,T}(), hyperedge_val::Union{UndefInitializer, E}=undef) where {T <: Real, V, E}
```
Comment (bkamins) [inline src/hypergraph.jl]: the same comment as for vertex function

File: src/hypergraph.jl [hunk 9]
```diff
@@ -108,41 +134,100 @@ Returns hyperedges for a given vertex `v_id` in a hypergraph `h`.
...
 end
 
+"""
+    set_vertex_value!(h::Hypergraph{T, V, E}, new_value::V, id::Int) where {T <: Real, V, E}
+
+Sets a new value `new_value` for the vertex `id` in the hypegraph `h`.
+
+"""
+function set_vertex_value!(h::Hypergraph{T, V, E}, new_value::V, id::Int) where {T <: Real, V, E}
+    @boundscheck checkbounds(h.vs, id)
```
Comment (bkamins) [inline src/hypergraph.jl]: you do not do `@inline` nor `@inbounds` below here so this is redundant.

File: src/hypergraph.jl [hunk 10]
```diff
@@ -108,41 +134,100 @@ Returns hyperedges for a given vertex `v_id` in a hypergraph `h`.
...
+end
+
+"""
+    get_vertex_value(h::Hypergraph{T, V, E}, id::Int) where {T <: Real, V, E}
+
+Returns a value stored at the vertex `id` in the hypergraph `h`.
+
+"""
+function get_vertex_value(h::Hypergraph{T, V, E}, id::Int) where {T <: Real, V, E}
+    @boundscheck checkbounds(h.vs, id)
```
Comment (bkamins) [inline src/hypergraph.jl]: you do not do `@inline` nor `@inbounds` below here so this is redundant.

File: src/hypergraph.jl [hunk 11]
```diff
@@ -108,41 +134,100 @@ Returns hyperedges for a given vertex `v_id` in a hypergraph `h`.
...
+end
+
+"""
+    get_hyperedge_value(h::Hypergraph{T, V, E}, id::Int) where {T <: Real, V, E}
+
+Returns a value stored at the hyperedge `id` in the hypergraph `h`.
+
+"""
+function get_hyperedge_value(h::Hypergraph{T, V, E}, id::Int) where {T <: Real, V, E}
+    @boundscheck checkbounds(h.hes, id)
```
Comment (bkamins) [inline src/hypergraph.jl]: you do not do `@inline` nor `@inbounds` below here so this is redundant.

File: src/hypergraph.jl (hunk 3 above)
Comment (pszufe) [reply src/hypergraph.jl]: or we can add two more parameters `v_default` and `e_default` and leave this decision for the user. Then they could write something like this (an we will need to document it):
```julia
h3 = Hypergraph{Float64,Union{String,Nothing}}(2,2; v_default=nothing)
```
The same pattern has been used for Julia `Array`s.

File: src/hypergraph.jl (hunk 6 above)
Comment (pszufe) [reply src/hypergraph.jl]: Again - this is the behavior from `Array`s - to be discussed

File: src/hypergraph.jl (hunk 3 above)
Comment (bkamins) [reply src/hypergraph.jl]: Whatever you prefer. For me having `nothing` as the default is a natural choice (this value exactly is meant to say that there is no value).

If you go the way you propose please remember that we have to make a union of `v_default` type and `V`.

File: src/hypergraph.jl (hunk 1 above)
Comment (bkamins) [reply src/hypergraph.jl]: I do not want to stick with `_meta`, but for me when I see `vs` it is like variables `k`, `l`, `m` in old Fortran code - I just do not know what they mean when I see them.

File: src/hypergraph.jl (hunk 6 above)
Comment (bkamins) [reply src/hypergraph.jl]: The behavior form `Array` has a different purpose. `undef` is introduced to avoid populating the array with initial values for performance reasons. It is used when you create thousands of arrays in hot code and populate their values based on some function, typically immediately after their creation. The reason is that e.g. `Vector{Float64, Nothing}` is less efficient than `Vector{Float64}` in linear algebra operations, so in such applications you might want to create an unitialized `Vector{Float64}`.

I would assume that our use case is different, as this metadata part is not a core of the functionality so it is better not to let `vs` etc. be filled with garbage. But I will not very strongly press here as long as everyone is clear what functionality we provide.

**Comment by bkamins** [comment]
Also with metadata added probably we should consider saving the information about the types of a `Hypergraph`, as now it might start becoming relevant. This also probably means that we should implement a binary format (even based on a serialize/deserialize I think should be enough for practical purposes).

**Comment by pszufe** [comment]
According to my tests the best option for graph serialization is the `BSON.jl`.
On one hand BSON format provides cross-language compatibility and on the other hand this library has a good stability.

**Comment by pszufe** [comment]
On other note - since serialization is currenty being developed by the Salerno team - I definitely propose to handle that in a separate pull request.

File: src/hypergraph.jl [hunk 12]
```diff
@@ -136,15 +136,18 @@ Returns hyperedges for a given vertex `v_id` in a hypergraph `h`.
 @inline gethyperedges(h::Hypergraph, v_id::Int) = h.v2he[v_id]
 
 """
-    add_vertex!(h::Hypergraph{T, V, E}; hyperedges::Dict{Int,T} = Dict{Int,T}(), vertex_meta::Union{V,Nothing} nothing undef) where {T <: Real, V, E}
+    add_vertex!(h::Hypergraph{T, V, E}; hyperedges::Dict{Int,T} = Dict{Int,T}(),
+        vertex_meta::Union{V,Nothing} nothing undef) where {T <: Real, V, E}
```
Comment (bkamins) [inline src/hypergraph.jl]: wrong line alignment. same below

**Comment by bkamins** [comment]
BSON.jl - OK. Serialization: who does work on serialization?

File: src/hypergraph.jl (hunk 12 above)
Comment (pszufe) [reply src/hypergraph.jl]: suggest the line alignment you want

**Comment by pszufe** [comment]
I understand that @spagnuolocarmine  and @aleant93  currently work on the hypergraph storage issues.

File: src/hypergraph.jl (hunk 12 above)
Comment (bkamins) [reply src/hypergraph.jl]: to `(`

**Comment by bkamins** [comment]
This is independent (they are working on comments). We do not have do do it in this PR, but then please open an Issue to fix it later.
